### PR TITLE
Replicate webhook pods for resiliency

### LINKED ISSUE #3386: Replicate webhook pods for resiliency
Currently we run one replica of the webhook deployment, which makes it a single point of failure in the event of a node outage. The job doesn't do anything that requires running as a single replica, so we should investigate running multiple replicas.

We can crib from Knative Serving, which configures:
* [a HorizontalPodAutoscaler (HPA) to run 1-5 replicas based on CPU usage](https://github.com/knative/serving/blob/c120748d8ca325621fe63f56a110871d2c6eaa0c/config/core/deployments/webhook-hpa.yaml#L23-L24)
* [a PodDisruptionBudget to ensure that at least 80% of replicas are available](https://github.com/knative/serving/blob/c120748d8ca325621fe63f56a110871d2c6eaa0c/config/core/deployments/webhook-hpa.yaml#L48)
* [pod anti-affinity to avoid having one node run all replicas](https://github.com/knative/serving/blob/c120748d8ca325621fe63f56a110871d2c6eaa0c/config/core/deployments/webhook.yaml#L37-L45)

/kind feature

### PR DESCRIPTION
**Pull request by raballew**
# Changes

Closes #3386

* Avoid node becoming single point of failure for webhook
* Ensure availability at node failure or high load
* Protect webhook in case of disruption

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Add HorizontalPodAutoscaler, PodDisruptionBudget and podAntiAffinity to webhook
```

### CODE REVIEW

**Comment by ImJasonH** [comment]
/kind feature
/ok-to-test

This looks good so far, let's see what the tests think. :) 

Can you squash your commits so this PR only contains one commit?

**Comment by afrittoli** [comment]
@raballew Thank you for your contribution!!
You (or your company) will need to sign the CLA - see link in above for instructions.

Could you include release notes in the PR description? 
It's in the [new PR template](https://raw.githubusercontent.com/tektoncd/pipeline/master/.github/pull_request_template.md) and it needs to be filled. Once you do so, the `do-not-merge/release-note-label-needed` will go away automatically.

Thank you again!

**Comment by raballew** [comment]
@afrittoli CLA signed
@ImJasonH commits are squashed now

Thanks for guiding me!

**Comment by afrittoli** [comment]
I've not really worked with the HPA before, so I tried this on my kind cluster.

After I got the metrics server running with:
```
          - --kubelet-insecure-tls
          - --kubelet-preferred-address-types=InternalIP
```

The HPA was still not happy, because of missing resource requests in the webhook pod:
```
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetResourceMetric  the HPA was unable to compute the replica count: missing request for cpu
```

After adding resources from in the deployment:
```
    resources:
      requests:
        cpu: 250m
```

if finally seems to be happy:
```
$ k get hpa -n tekton-pipelines
NAME                       REFERENCE                             TARGETS   MINPODS   MAXPODS   REPLICAS   AGE
tekton-pipelines-webhook   Deployment/tekton-pipelines-webhook   3%/100%   1         5         1          135m

$ k describe hpa -n tekton-pipelines
Name:                                                  tekton-pipelines-webhook
Namespace:                                             tekton-pipelines
Labels:                                                app.kubernetes.io/component=webhook
                                                       app.kubernetes.io/instance=default
                                                       app.kubernetes.io/name=webhook
                                                       app.kubernetes.io/part-of=tekton-pipelines
                                                       app.kubernetes.io/version=devel
                                                       pipeline.tekton.dev/release=devel
                                                       version=devel
Annotations:                                           <none>
CreationTimestamp:                                     Fri, 16 Oct 2020 08:23:24 +0100
Reference:                                             Deployment/tekton-pipelines-webhook
Metrics:                                               ( current / target )
  resource cpu on pods  (as a percentage of request):  3% (8m) / 100%
Min replicas:                                          1
Max replicas:                                          5
Deployment pods:                                       1 current / 1 desired
Conditions:
  Type            Status  Reason              Message
  ----            ------  ------              -------
  AbleToScale     True    ReadyForNewScale    recommended size matches current size
  ScalingActive   True    ValidMetricFound    the HPA was able to successfully calculate a replica count from cpu resource utilization (percentage of request)
  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
```

According to the [HPA docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/):
```
Please note that if some of the Pod's containers do not have the relevant resource request set, CPU utilization for the Pod will not be defined and the autoscaler will not take any action for that metric. See the algorithm details section below for more information about how the autoscaling algorithm works.
```

So I think that for this to work we need to set resource requests too.
The good news is that without a metrics service the HPA will be un-happy, but the deployment will still work fine, so for dev environments this is pretty much transparent. So I think it should be ok to have this the "default" deployment.

**Review by afrittoli — COMMENTED** [review]
Thanks for this. I think we should clarify if we need to add resources.
TBH I don't know what a good value for CPU would be? @ImJasonH wdyt?

File: config/webhook-hpa.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2019 The Tekton Authors
```
[the remark this answers is not in the document]
Comment (raballew) [reply config/webhook-hpa.yaml:1]: Sure I will bump the year.

@afrittoli I also looked through all [resource definitions](https://github.com/tektoncd/pipeline/tree/master/config) and most of the files were modified through the course of this year and still use 2019 copyright year. If I am not wrong, technically the copyright year should be updated when you made contributions to a file within that year.

File: config/webhook-hpa.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/name: webhook
+    app.kubernetes.io/component: webhook
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/version: "devel"
+    app.kubernetes.io/part-of: tekton-pipelines
+    # tekton.dev/release value replaced with inputs.params.versionTag in pipeline/tekton/publish.yaml
+    pipeline.tekton.dev/release: "devel"
+    # labels below are related to istio and should not be used for resource lookup
```
[the remark this answers is not in the document]
Comment (raballew) [reply config/webhook-hpa.yaml:56]: I simply used the same values as for the [deployment](https://github.com/tektoncd/pipeline/blob/d87692381071d27e2c10d40f63c85625803f0db6/config/webhook.yaml#L28) and [service](https://github.com/tektoncd/pipeline/blob/d87692381071d27e2c10d40f63c85625803f0db6/config/webhook.yaml#L116) as the HPA is part of the webhook and therefore should use the same set of labels.

File: config/webhook-hpa.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (afrittoli) [reply config/webhook-hpa.yaml]: Uhm, the practise that I've always followed and that I've seen being used in several open source projects is it to set the year to that of the creation of the file, and do not bother updating the date again.
There are a lot of different opinions about this in the internet. The feeling I get after reading a few opinions is that the year is not so critical in the copyright notice;  however company lawyers usually insist on having it.

We could add guidance about this in the community repo @vdemeester @bobcatfish @ImJasonH @abayer

**Comment by ImJasonH** [comment]
> Thanks for this. I think we should clarify if we need to add resources.
> 
> TBH I don't know what a good value for CPU would be? @ImJasonH wdyt?

Two options:

1. Run a Tekton installation on your cluster and give it some load, and observe how many resources it uses, then give it 2x or so to give us some headroom.
1. Crib from Knative even further, and copy their resource requests: https://github.com/knative/serving/blob/master/config/core/deployments/webhook.yaml

I'd probably just borrow from Knative, I don't think either of us do anything very resource intensive in our webhook checks, and if it turns out to be too low we can bump it up in the future.

**Comment by afrittoli** [comment]
> 2\. https://github.com/knative/serving/blob/master/config/core/deployments/webhook.yaml

ok, (2) sounds reasonable to me, also 100m seems small enough for a validation service.
Do we needs limits? If the limit is reached, the HPA will spawn a new instance?

```
        resources:
          requests:
            cpu: 100m
            memory: 100Mi
          limits:
            cpu: 500m
            memory: 500Mi
```

**Comment by raballew** [comment]
> If the limit is reached, the HPA will spawn a new instance?

According to [HPA docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/):

```
 [...] the controller calculates the utilization value as a percentage of the equivalent resource request on the containers in each Pod.
```

When a new replica should be spawned, is described [here](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details).

**Comment by afrittoli** [comment]
Thanks for the updates! 

This looks good to me know, the only thing missing is documentation, but I'm ok doing that as a separate PR if you prefer.
We need to document that a metrics server is required for this to work, and in general it would be good to have an HA setup dedicated document or paragraph.

CC'ing @qu1queee since he's been looking into HA documentation for Tekton.

/lgtm

**Comment by raballew** [comment]
@afrittoli I will do it in this PR, so everything related to webhooks and HPA is bundled here. Could you point me to the correct file where to add the documentation?

**Comment by afrittoli** [comment]
I think the entry point might be the [install and configuration doc](https://github.com/tektoncd/pipeline/blob/17b69e2d823893f778aa9fd2588e8679bcd1c113/docs/install.md). 
It could be a chapter in that page, the content could be in the install page directly or perhaps in a dedicated `ha.md` with the HA concept and how to configure it.

If you decide to go for a dedicated new file, could you also file a PR to the [website repo](https://github.com/tektoncd/website/pull/179), so that it shows up on [tekton.dev](https://tekton.dev) too?

File: config/webhook-hpa.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlorenc) [reply config/webhook-hpa.yaml]: In general the guidance I've seen is to bump the year if you're touching the file. Bulk commits to update the year everywhere aren't great for history/blame tracking, but keeping it up to date is slightly better than not...

**Comment by raballew** [comment]
@afrittoli I have added the HA docs to the install page as the amount of documentation needed for this topic is rather small. Once more components such as the [controller](https://github.com/tektoncd/pipeline/issues/3404) implement a scaling mechanism as well, this section should be moved to a dedicated `ha.md` file.

**Review by afrittoli — COMMENTED** [review]
Thanks for the docs on this!
/approve

**Pull request merged**